{
	"title": "Chapter 13: Concurrency",
	"questions": [
    	{
        	"number": 1,
        	"title": "Given the following code snippet, which options correctly create a parallel stream? (Choose all that apply.)",
        	"code": "var c = new ArrayList<Thread>();\nvar s = c.stream();\nvar p = _________________;",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "new ParallelStream(s)" },
            	{ "id": "B", "text": "c.parallel()" },
            	{ "id": "C", "text": "s.parallelStream()" },
            	{ "id": "D", "text": "c.parallelStream()" },
            	{ "id": "E", "text": "new ParallelStream(c)" },
            	{ "id": "F", "text": "s.parallel()" }
            ],
        	"answer": {
            	"ids": ["F","D"],
            	"explanation": "There is no such class within the Java API called ParallelStream, so options A and E are incorrect. The method defined in the Stream class to create a parallel stream from an existing stream is parallel(); therefore, option F is correct, and option C is incorrect. The method defined in the Collection class to create a parallel stream from a collection is parallelStream(); therefore, option D is correct, and option B is incorrect."
            }
        },
    	{
        	"number": 2,
        	"title": "Given that the sum of the numbers from 1 (inclusive) to 10 (exclusive) is 45, what are the possible results of executing the following program? (Choose all that apply.)",
        	"code": "1:  import java.util.concurrent.locks.*;\n2:  import java.util.stream.*;\n3:  public class Bank {\n4:     private Lock vault = new ReentrantLock();\n5:     private int total = 0;\n6:     public void deposit(int value) {\n7:        try {\n8:           vault.tryLock();\n9:           total += value;\n10:       } finally { vault.unlock(); }\n11:    }\n12:    public static void main(String[] unused) {\n13:       var bank = new Bank();\n14:       IntStream.range(1, 10).parallel()\n15:          .forEach(s -> bank.deposit(s));\n16:       System.out.println(bank.total);\n17:    } }",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "45 is printed" },
            	{ "id": "B", "text": "A number less than 45 is printed" },
            	{ "id": "C", "text": "A number greater than 45 is printed" },
            	{ "id": "D", "text": "An exception is thrown" },
            	{ "id": "E", "text": "None of the above, as the code does not compile" }
            ],
        	"answer": {
            	"ids": ["D","A"],
            	"explanation": "The tryLock() method returns immediately with a value of false if the lock cannot be acquired. Unlike lock(), it does not wait for a lock to become available. This code fails to check the return value on line 8, resulting in the protected code being entered regardless of whether the lock is obtained. In some executions (when tryLock() returns true on every call), the code will complete successfully and print 45 at runtime, making option A correct. On other executions (when tryLock() returns false at least once), the unlock() method on line 10 will throw an IllegalMonitorStateException at runtime, making option D correct. Option B would be possible if line 10 did not throw an exception."
            }
        },
    	{
        	"number": 3,
        	"title": "Which of the following statements about the Callable call() and Runnable run() methods are correct? (Choose all that apply.)",
        	"code": null,
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "Both methods return void" },
            	{ "id": "B", "text": "Both can throw unchecked exceptions" },
            	{ "id": "C", "text": "Both can be implemented with lambda expressions" },
            	{ "id": "D", "text": "Runnable returns a generic type" },
            	{ "id": "E", "text": "Both can throw checked exceptions" },
            	{ "id": "F", "text": "Callable returns a generic type" }
            ],
        	"answer": {
            	"ids": ["B","F","C"],
            	"explanation": "Runnable returns void and Callable returns a generic type, making options A and D incorrect and option F correct. All methods are capable of throwing unchecked exceptions, so option B is correct. Only Callable is capable of throwing checked exceptions, so option E is incorrect. Both Runnable and Callable are functional interfaces that can be implemented with a lambda expression, so option C is also correct."
            }
        },
    	{
        	"number": 4,
        	"title": "Which lines need to be changed to make the code compile? (Choose all that apply.)",
        	"code": "ExecutorService service =   // w1\n   Executors.newSingleThreadScheduledExecutor();\nservice.scheduleWithFixedDelay(() -> {\n   System.out.println(\"Open Zoo\");\n   return null;   // w2\n}, 0, 1, TimeUnit.MINUTES);\nvar result = service.submit(() ->   // w3\n   System.out.println(\"Wake Staff\"));\nSystem.out.println(result.get());   // w4",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "It compiles and runs without issue" },
            	{ "id": "B", "text": "Line w1" },
            	{ "id": "C", "text": "Line w2" },
            	{ "id": "D", "text": "Line w3" },
            	{ "id": "E", "text": "Line w4" },
            	{ "id": "F", "text": "It compiles but throws an exception at runtime" }
            ],
        	"answer": {
            	"ids": ["B","C"],
            	"explanation": "The code does not compile, so options A and F are incorrect. The first problem is that although a ScheduledExecutorService is created, it is assigned to an ExecutorService. The type of the variable on line w1 would have to be updated to ScheduledExecutorService for the code to compile, making option B correct. The second problem is that scheduleWithFixedDelay() supports only Runnable, not Callable, and any attempt to return a value is invalid in a Runnable lambda expression; therefore, line w2 will also not compile, and option C is correct. The rest of the lines compile without issue, so options D and E are incorrect."
            }
        },
    	{
        	"number": 5,
        	"title": "What statement about the following code is true?",
        	"code": "var value1 = new AtomicLong(0);\nfinal long[] value2 = {0};\nIntStream.iterate(1, i -> 1).limit(100).parallel()\n   .forEach(i -> value1.incrementAndGet());\nIntStream.iterate(1, i -> 1).limit(100).parallel()\n   .forEach(i -> ++value2[0]);\nSystem.out.println(value1+\" \"+value2[0]);",
        	"multipleOptions": false,
        	"options": [
            	{ "id": "A", "text": "It outputs 100 100" },
            	{ "id": "B", "text": "It outputs 100 99" },
            	{ "id": "C", "text": "The output cannot be determined ahead of time" },
            	{ "id": "D", "text": "The code does not compile" },
            	{ "id": "E", "text": "It compiles but throws an exception at runtime" },
            	{ "id": "F", "text": "It compiles but enters an infinite loop at runtime" },
            	{ "id": "G", "text": "None of the above" }
            ],
        	"answer": {
            	"ids": ["C"],
            	"explanation": "The code compiles and runs without throwing an exception or entering an infinite loop, so options D, E, and F are incorrect. The key here is that the increment operator ++ is not atomic. While the first part of the output will always be 100, the second part is nondeterministic. It may output any value from 1 to 100, because the threads can overwrite each other's work. Therefore, option C is the correct answer, and options A and B are incorrect."
            }
        },
    	{
        	"number": 6,
        	"title": "Which statements about the following code are correct? (Choose all that apply.)",
        	"code": "var data = List.of(2,5,1,9,8);\ndata.stream().parallel()   .mapToInt(s -> s)\n   .peek(System.out::print)\n   .forEachOrdered(System.out::print);",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "The peek() method will print the entries in the sorted order: 12589" },
            	{ "id": "B", "text": "The peek() method will print the entries in the original order: 25198" },
            	{ "id": "C", "text": "The peek() method will print the entries in an order that cannot be determined ahead of time" },
            	{ "id": "D", "text": "The forEachOrdered() method will print the entries in the sorted order: 12589" },
            	{ "id": "E", "text": "The forEachOrdered() method will print the entries in the original order: 25198" },
            	{ "id": "F", "text": "The forEachOrdered() method will print the entries in an order that cannot be determined ahead of time" },
            	{ "id": "G", "text": "The code does not compile" }
            ],
        	"answer": {
            	"ids": ["C","E"],
            	"explanation": "The code compiles, so option G is incorrect. The peek() method on a parallel stream will process the elements concurrently, so the order cannot be determined ahead of time, and option C is correct. The forEachOrdered() method will process the elements in the order in which they are stored in the stream, making option E correct. None of the methods sort the elements, so options A and D are incorrect."
            }
        },
    	{
        	"number": 7,
        	"title": "Fill in the blanks: __________ occur(s) when two or more threads are blocked forever but both appear active. _______ occur(s) when two or more threads try to complete a related task at the same time, resulting in invalid or unexpected data",
        	"code": null,
        	"multipleOptions": false,
        	"options": [
            	{ "id": "A", "text": "Livelock, Deadlock" },
            	{ "id": "B", "text": "Deadlock, Starvation" },
            	{ "id": "C", "text": "Race conditions, Deadlock" },
            	{ "id": "D", "text": "Livelock, Race conditions" },
            	{ "id": "E", "text": "Starvation, Race conditions" },
            	{ "id": "F", "text": "Deadlock, Livelock" }
            ],
        	"answer": {
            	"ids": ["D"],
            	"explanation": "Livelock occurs when two or more threads are conceptually blocked forever,although they are each still active and trying to complete their task. A race condition is an undesirable result that occurs when two tasks that should have been completed sequentially are completed at the same time. For these reasons, option D is correct."
            }
        },
    	{
        	"number": 8,
        	"title": "Assuming this class is accessed by only a single thread at a time, what is the result of calling the countIceCreamFlavors() method?",
        	"code": "import java.util.stream.LongStream;\npublic class Flavors {\n   private static int counter;\n   public static void countIceCreamFlavors()  {\n      counter = 0;\n      Runnable task = () -> counter++;\n      LongStream.range(0, 500)\n         .forEach(m -> new Thread(task).run());\n      System.out.println(counter);\n   } }",
        	"multipleOptions": false,
        	"options": [
            	{ "id": "A", "text": "The method consistently prints a number less than 500" },
            	{ "id": "B", "text": "The method consistently prints 500" },
            	{ "id": "C", "text": "The method compiles and prints a value, but that value cannot be determined ahead of time" },
            	{ "id": "D", "text": "The method does not compile" },
            	{ "id": "E", "text": "The method compiles but throws an exception at runtime" },
            	{ "id": "F", "text": "None of the above" }
            ],
        	"answer": {
            	"ids": ["B"],
            	"explanation": "Be wary of run() vs. start() on the exam! The method looks like it executes a task concurrently, but it runs synchronously. In each iteration of the forEach() loop, the process waits for the run() method to complete before moving on. For this reason, the code is thread-safe. Since the program consistently prints 500 at runtime, option B is correct. Note that if start() had been used instead of run() (or the stream was parallel), then the output would be indeterminate, and option C would have been correct."
            }
        },
    	{
        	"number": 9,
        	"title": "Which happens when a new task is submitted to an ExecutorService in which no threads are available?",
        	"code": null,
        	"multipleOptions": false,
        	"options": [
            	{ "id": "A", "text": "The executor throws an exception when the task is submitted" },
            	{ "id": "B", "text": "The executor discards the task without completing it" },
            	{ "id": "C", "text": "The executor adds the task to an internal queue and completes when there is an available thread" },
            	{ "id": "D", "text": "The thread submitting the task waits on the submit call until a thread is available before continuing" },
            	{ "id": "E", "text": "The executor stops an existing task and starts the newly submitted one" }
            ],
        	"answer": {
            	"ids": ["C"],
            	"explanation": "If a task is submitted to a thread executor, and the thread executor does not have any available threads, the call to the task will return immediately with the task being queued internally by the thread executor. For this reason, option C is the correct answer."
            }
        },
    	{
        	"number": 10,
        	"title": "What is the result of executing the following code snippet?",
        	"code": "List<Integer> lions = new ArrayList<>(List.of(1,2,3));\nList<Integer> tigers = new CopyOnWriteArrayList<>(lions);\nSet<Integer> bears = new ConcurrentSkipListSet<>();\nbears.addAll(lions);\nfor(Integer item: tigers) tigers.add(4); // x1\nfor(Integer item: bears) bears.add(5);   // x2\nSystem.out.println(lions.size() + \" \" + tigers.size()\n   + \" \" + bears.size());",
        	"multipleOptions": false,
        	"options": [
            	{ "id": "A", "text": "It outputs 3 6 4" },
            	{ "id": "B", "text": "It outputs 6 6 6" },
            	{ "id": "C", "text": "It outputs 6 3 4" },
            	{ "id": "D", "text": "The code does not compile" },
            	{ "id": "E", "text": "It compiles but throws an exception at runtime on line x1" },
            	{ "id": "F", "text": "It compiles but throws an exception at runtime on line x2" },
            	{ "id": "G", "text": "It compiles but enters an infinite loop at runtime" }
            ],
        	"answer": {
            	"ids": ["A"],
            	"explanation": "The code compiles without issue, so option D is incorrect. The CopyOnWriteArrrayList class is designed to preserve the original list on iteration, so the first loop will be executed exactly three times and, in the process, will increase the size of tigers to six elements. The ConcurrentSkipListSet class allows modifications, and since it enforces the uniqueness of its elements, the value 5 is added only once, leading to a total of four elements in bears. Finally, despite using the elements of lions to populate the collections, tigers and bears are not backed by the original list, so the size of lions is 3 throughout this program. For these reasons, the program prints 3 6 4, and option A is correct."
            }
        },
    	{
        	"number": 11,
        	"title": "What statements about the following code are true? (Choose all that apply.)",
        	"code": "Integer i1 = List.of(1, 2, 3, 4, 5).stream().findAny().get();\nsynchronized(i1) { // y1\n   Integer i2 = List.of(6, 7, 8, 9, 10)\n      .parallelStream()\n      .sorted()\n      .findAny().get(); // y2\n   System.out.println(i1 + \" \" + i2);\n}",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "The first value printed is always 1" },
            	{ "id": "B", "text": "The second value printed is always 6" },
            	{ "id": "C", "text": "The code will not compile because of line y1" },
            	{ "id": "D", "text": "The code will not compile because of line y2" },
            	{ "id": "E", "text": "The code compiles but throws an exception at runtime" },
            	{ "id": "F", "text": "The output cannot be determined ahead of time" },
            	{ "id": "G", "text": "It compiles but waits forever at runtime" }
            ],
        	"answer": {
            	"ids": ["F"],
            	"explanation": "The code compiles and runs without issue, so options C, D, E, and G are incorrect. There are two important things to notice. First, synchronizing on the first variable doesn't impact the results of the code. Second, sorting on a parallel stream does not mean that findAny() will return the first record. The findAny() method will return the value from the first thread that retrieves a record. Therefore, the output is not guaranteed, and option F is correct. Option A looks correct, but even on serial streams, findAny() is free to select any element."
            }
        },
    	{
        	"number": 12,
        	"title": "Assuming each call to takeNap() takes five seconds to execute without throwing an exception, what is the expected result of executing the following code snippet?",
        	"code": "ExecutorService service = Executors.newFixedThreadPool(4);\ntry {\n   service.execute(() -> takeNap());\n   service.execute(() -> takeNap());\n   service.execute(() -> takeNap());\n} finally {\n   service.shutdown();\n}\nservice.awaitTermination(2, TimeUnit.SECONDS);\nSystem.out.println(\"DONE!\");",
        	"multipleOptions": false,
        	"options": [
            	{ "id": "A", "text": "It will immediately print DONE!" },
            	{ "id": "B", "text": "It will pause for 2 seconds and then print DONE!" },
            	{ "id": "C", "text": "It will pause for 5 seconds and then print DONE!" },
            	{ "id": "D", "text": "It will pause for 15 seconds and then print DONE!" },
            	{ "id": "E", "text": "It will throw an exception at runtime" },
            	{ "id": "F", "text": "None of the above, as the code does not compile" }
            ],
        	"answer": {
            	"ids": ["B"],
            	"explanation": "The code snippet submits three tasks to an ExecutorService, shuts it down, and then waits for the results. The awaitTermination() method waits a specified amount of time for all tasks to complete and the service to finish shutting down. Since each five-second task is still executing, the awaitTermination() method will return with a value of false after two seconds but not throw an exception. For these reasons, option B is correct."
            }
        },
    	{
        	"number": 13,
        	"title": "What statements about the following code are true? (Choose all that apply.)",
        	"code": "System.out.print(List.of(\"duck\",\"flamingo\",\"pelican\")\n   .parallelStream().parallel()   // q1\n   .reduce(0,\n      (c1, c2) -> c1.length() + c2.length(),  // q2\n      (s1, s2) -> s1 + s2));      // q3",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "It compiles and runs without issue, outputting the total length of all strings in the stream" },
            	{ "id": "B", "text": "The code will not compile because of line q1" },
            	{ "id": "C", "text": "The code will not compile because of line q2" },
            	{ "id": "D", "text": "The code will not compile because of line q3" },
            	{ "id": "E", "text": "It compiles but throws an exception at runtime" },
            	{ "id": "F", "text": "None of the above" }
            ],
        	"answer": {
            	"ids": ["C"],
            	"explanation": "The code does not compile, so options A and E are incorrect. The problem here is that c1 is an Integer and c2 is a String, so the code fails to combine on line q2, since calling length() on an Integer is not allowed, and option C is correct. The rest of the lines compile without issue. Note that calling parallel() on an already parallel stream is allowed, and it may return the same object."
            }
        },
    	{
        	"number": 14,
        	"title": "What statements about the following code snippet are true? (Choose all that apply.)",
        	"code": "Object o1 = new Object();\nObject o2 = new Object();\nvar service = Executors.newFixedThreadPool(2);\nvar f1 = service.submit(() -> {\n   synchronized (o1) {\n      synchronized (o2) { System.out.print(\"Tortoise\"); }   }\n});\nvar f2 = service.submit(() -> {\n   synchronized (o2) {\n      synchronized (o1) { System.out.print(\"Hare\"); }\n   }\n});\nf1.get();\nf2.get();",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "The code will always output Tortoise followed by Hare" },
            	{ "id": "B", "text": "The code will always output Hare followed by Tortoise" },
            	{ "id": "C", "text": "If the code does output anything, the order cannot be determined" },
            	{ "id": "D", "text": "The code does not compile" },
            	{ "id": "E", "text": "The code compiles but may produce a deadlock at runtime" },
            	{ "id": "F", "text": "The code compiles but may produce a livelock at runtime" },
            	{ "id": "G", "text": "It compiles but throws an exception at runtime" }
            ],
        	"answer": {
            	"ids": ["C","E"],
            	"explanation": "The code compiles without issue, so option D is incorrect. Since both tasks are submitted to the same thread executor pool, the order cannot be determined, so options A and B are incorrect, and option C is correct. The key here is that the order in which the resources o1 and o2 are synchronized could result in a deadlock. For example, if the first thread gets a lock on o1 and the second thread gets a lock on o2 before either thread can get their second lock, the code will hang at runtime, making option E correct. The code cannot produce a livelock, since both threads are waiting, so option F is incorrect. Finally, if a deadlock does occur, an exception will not be thrown, so option G is incorrect."
            }
        },
    	{
        	"number": 15,
        	"title": "Which statement about the following code snippet is correct?",
        	"code": "2: var cats = Stream.of(\"leopard\", \"lynx\", \"ocelot\", \"puma\")\n3:    .parallel();\n4: var bears = Stream.of(\"panda\",\"grizzly\",\"polar\").parallel();\n5: var data = Stream.of(cats,bears).flatMap(s -> s)\n6:    .collect(Collectors.groupingByConcurrent(\n7:       s -> !s.startsWith(\"p\")));\n8: System.out.println(data.get(false).size()\n9:    + \" \" + data.get(true).size());",
        	"multipleOptions": false,
        	"options": [
            	{ "id": "A", "text": "It outputs 3 4" },
            	{ "id": "B", "text": "It outputs 4 3" },
            	{ "id": "C", "text": "The code will not compile because of line 6" },
            	{ "id": "D", "text": "The code will not compile because of line 7" },
            	{ "id": "E", "text": "The code will not compile because of line 8" },
            	{ "id": "F", "text": "It compiles but throws an exception at runtime" }
            ],
        	"answer": {
            	"ids": ["A"],
            	"explanation": "The code compiles and runs without issue, so options C, D, E, and F are incorrect. The collect() operation groups the animals into those that do and do not start with the letter p. Note that there are four animals that do not start with the letter p and three animals that do. The logical complement operator (!) before the startsWith() method means that results are reversed, so the output is 3 4, and option A is correct, making option B incorrect."
            }
        },
    	{
        	"number": 16,
        	"title": "Assuming one minute is enough time for all the threads within this program to complete, what are the possible results of executing the following program? (Choose all that apply.)",
        	"code": "public class RocketShip {\n   private volatile int fuel;\n   private void launch(int checks) {\n      var p = new ArrayList<Thread>();\n      for (int i = 0; i < checks; i++)\n         p.add(new Thread(() -> fuel++));\n      p.forEach(Thread::interrupt);\n      p.forEach(Thread::start);\n      p.forEach(Thread::interrupt);\n   }\n   public static void main(String[] args) throws Exception {\n      var ship = new RocketShip();\n      ship.launch(100);\n      Thread.sleep(60*1000);\n      System.out.print(ship.fuel);\n   }\n}",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "It prints a number less than 100" },
            	{ "id": "B", "text": "It prints 100" },
            	{ "id": "C", "text": "It prints a number greater than 100" },
            	{ "id": "D", "text": "It does not compile" },
            	{ "id": "E", "text": "It compiles but throws an InterruptedException at runtime" }
            ],
        	"answer": {
            	"ids": ["A","B"],
            	"explanation": "The code compiles just fine. If the calls to fuel++ are ordered sequentially, then the program will print 100 at runtime, making option B correct. On the other hand, the calls may overwrite each other. The volatile attribute only guarantees memory consistency, not thread-safety, making option A correct and option C incorrect. Option E is also incorrect, as no InterruptedException is thrown by this code. Remember, interrupt() only impacts a thread that is in a WAITING or TIMED_WAITING state. Calling interrupt() on a thread in a NEW or RUNNABLE state has no impact unless the code is running and explicitly checking the isInterrupted() method."
            }
        },
    	{
        	"number": 17,
        	"title": "Which statements about methods in ReentrantLock are correct? (Choose all that apply.)",
        	"code": null,
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "The lock() method will attempt to acquire a lock without waiting indefinitely for it" },
            	{ "id": "B", "text": "The testLock() method will attempt to acquire a lock without waiting indefinitely for it" },
            	{ "id": "C", "text": "The attemptLock() method will attempt to acquire a lock without waiting indefinitely for it" },
            	{ "id": "D", "text": "By default, a ReentrantLock fairly releases to each thread in the order in which it was requested" },
            	{ "id": "E", "text": "Calling the unlock() method once will release a resource so that other threads can obtain the lock" },
            	{ "id": "F", "text": "None of the above" }
            ],
        	"answer": {
            	"ids": ["F"],
            	"explanation": "The lock() method will wait indefinitely for a lock, so option A is incorrect. Options B and C are also incorrect, as the correct method name to attempt to acquire a lock is tryLock(). Option D is incorrect, as fairness is set to false by default and must be enabled by using an overloaded constructor. Finally, option E is incorrect because a thread that holds the lock may have called lock() or tryLock() multiple times. A thread needs to call unlock() once for each call to lock() and successful tryLock(). Option F is the correct answer since none of the other options are valid statements."
            }
        },
    	{
        	"number": 18,
        	"title": "Which of the following are valid Callable expressions? (Choose all that apply.)",
        	"code": null,
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "a -> {return 10;}" },
            	{ "id": "B", "text": "() -> {String s = \"\";}" },
            	{ "id": "C", "text": "() -> 5" },
            	{ "id": "D", "text": "() -> {return null}" },
            	{ "id": "E", "text": "() -> \"The\" + \"Zoo\"" },
            	{ "id": "F", "text": "(int count) -> count+1" },
            	{ "id": "G", "text": "() -> {System.out.println(\"Giraffe\"); return 10;}" }
            ],
        	"answer": {
            	"ids": ["C","G","E"],
            	"explanation": "A Callable lambda expression takes no values and returns a generic type; therefore, options C, E, and G are correct. Options A and F are incorrect because they both take an input parameter. Option B is incorrect because it does not return a value. Option D is not a valid lambda expression, because it is missing a semicolon at the end of the return statement, which is required when inside braces {}."
            }
        },
    	{
        	"number": 19,
        	"title": "What is the result of executing the following application? (Choose all that apply.)",
        	"code": "import java.util.concurrent.*;\nimport java.util.stream.*;\npublic class PrintConstants {\n   public static void main(String[] args) {\n      var s = Executors.newScheduledThreadPool(10);\n      DoubleStream.of(3.14159,2.71828)   // b1\n         .forEach(c -> s.submit(         // b2\n            () -> System.out.println(10*c)));  // b3\n      s.execute(() -> System.out.println(\"Printed\"));\n   } }",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "It compiles and outputs the two numbers followed by Printed" },
            	{ "id": "B", "text": "The code will not compile because of line b1" },
            	{ "id": "C", "text": "The code will not compile because of line b2" },
            	{ "id": "D", "text": "The code will not compile because of line b3" },
            	{ "id": "E", "text": "It compiles, but the output cannot be determined ahead of time" },
            	{ "id": "F", "text": "It compiles but throws an exception at runtime" },
            	{ "id": "G", "text": "It compiles but waits forever at runtime" }
            ],
        	"answer": {
            	"ids": ["G","E"],
            	"explanation": "The application compiles and does not throw an exception. Even though the stream is processed in sequential order, the tasks are submitted to a thread executor, which may complete the tasks in any order. Therefore, the output cannot be determined ahead of time, and option E is correct. Finally, the thread executor is never shut down; therefore, the code will run but never terminate, making option G also correct."
            }
        },
    	{
        	"number": 20,
        	"title": "What is the result of executing the following program? (Choose all that apply.)",
        	"code": "import java.util.*;\nimport java.util.concurrent.*;\nimport java.util.stream.*;\npublic class PrintCounter {\n   static int count = 0;\n   public static void main(String[] args) throws\n                     InterruptedException, ExecutionException {\n      var service = Executors.newSingleThreadExecutor();\n      try {\n         var r = new ArrayList<Future<?>>();\n         IntStream.iterate(0,i -> i+1).limit(5).forEach(\n            i -> r.add(service.execute(() -> {count++;})) // n1\n         );\n         for(Future<?> result : r) {\n            System.out.print(result.get()+\" \"); // n2\n         }\n      } finally { service.shutdown(); }\n   } }",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "It prints 0 1 2 3 4" },
            	{ "id": "B", "text": "It prints 1 2 3 4 5" },
            	{ "id": "C", "text": "It prints null null null null null" },
            	{ "id": "D", "text": "It hangs indefinitely at runtime" },
            	{ "id": "E", "text": "The output cannot be determined" },
            	{ "id": "F", "text": "The code will not compile because of line n1" },
            	{ "id": "G", "text": "The code will not compile because of line n2" }
            ],
        	"answer": {
            	"ids": ["F"],
            	"explanation": "The key to solving this question is to remember that the execute() method returns void, not a Future object. Therefore, line n1 does not compile, and option F is the correct answer. If the submit() method had been used instead of execute(), option C would have been the correct answer, as the output of the submit(Runnable) task is a Future<?> object that can only return null on its get() method."
            }
        },
    	{
        	"number": 21,
        	"title": "Given the following code snippet and blank lines on p1 and p2, which values guarantee that 1 is printed at runtime? (Choose all that apply.)",
        	"code": "var data = List.of(List.of(1,2),\n   List.of(3,4),\n   List.of(5,6));\ndata.___________    // p1   .flatMap(s -> s.stream())\n   .____________    // p2\n   .ifPresent(System.out::print);",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "stream() on line p1, findFirst() on line p2" },
            	{ "id": "B", "text": "stream() on line p1, findAny() on line p2" },
            	{ "id": "C", "text": "parallelStream() on line p1, findAny() on line p2" },
            	{ "id": "D", "text": "parallelStream() on line p1, findFirst() on line p2" },
            	{ "id": "E", "text": "The code does not compile regardless of what is inserted into the blanks" },
            	{ "id": "F", "text": "None of the above" }
            ],
        	"answer": {
            	"ids": ["D","A"],
            	"explanation": "The findFirst() method guarantees the first element in the stream will be returned, whether it is serial or parallel, making options A and D correct. While option B may consistently print 1 at runtime, the behavior of findAny() on a serial stream is not guaranteed, so option B is incorrect. Option C is likewise incorrect, with the output being random at runtime."
            }
        },
    	{
        	"number": 22,
        	"title": "Assuming one minute is enough time for the tasks submitted to the service executor to complete, what is the result of executing countSheep()? (Choose all that apply.)",
        	"code": "import java.util.concurrent.*;\nimport java.util.concurrent.atomic.*;\npublic class BedTime {\n   private AtomicInteger s1 = new AtomicInteger(0); // w1\n   private int s2 = 0;\n \n   private void countSheep() throws InterruptedException {\n      var service = Executors.newSingleThreadExecutor(); // w2\n      try {\n         for (int i = 0; i < 100; i++)\n         service.execute(() -> {\n            s1.getAndIncrement(); s2++; }); // w3\n         Thread.sleep(60*1000);\n         System.out.println(s1 + \" \" + s2);\n      } finally { service.shutdown(); }\n   }\n   public static void main(String… nap) throws InterruptedException {\n      new BedTime().countSheep();\n   } }",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "The method consistently prints 100 99" },
            	{ "id": "B", "text": "The method consistently prints 100 100" },
            	{ "id": "C", "text": "The output cannot be determined ahead of time" },
            	{ "id": "D", "text": "The code will not compile because of line w1" },
            	{ "id": "E", "text": "The code will not compile because of line w2" },
            	{ "id": "F", "text": "The code will not compile because of line w3" },
            	{ "id": "G", "text": "It compiles but throws an exception at runtime" }
            ],
        	"answer": {
            	"ids": ["B"],
            	"explanation": "The code compiles and runs without issue. The key aspect to notice in the code is that a single-thread executor is used, meaning that no task will be executed concurrently. Therefore, the results are valid and predictable, with 100 100 being the output, and option B is the correct answer. If a thread executor with more threads was used, then the s2++ operations could overwrite each other, making the second value indeterminate at the end of the program. In this case, option C would be the correct answer."
            }
        },
    	{
        	"number": 23,
        	"title": "What is the result of executing the following application? (Choose all that apply.)",
        	"code": "import java.util.concurrent.*;\nimport java.util.stream.*;\npublic class StockRoomTracker {\n   public static void await(CyclicBarrier cb) { // j1\n      try { cb.await(); } catch (Exception e) {}\n   }   public static void main(String[] args) {\n      var cb = new CyclicBarrier(10,\n         () -> System.out.println(\"Stock Room Full!\")); // j2\n      IntStream.iterate(1, i -> 1).limit(9).parallel()\n         .forEach(i -> await(cb)); // j3\n   } }",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "It outputs Stock Room Full!" },
            	{ "id": "B", "text": "The code will not compile because of line j1" },
            	{ "id": "C", "text": "The code will not compile because of line j2" },
            	{ "id": "D", "text": "The code will not compile because of line j3" },
            	{ "id": "E", "text": "It compiles but throws an exception at runtime" },
            	{ "id": "F", "text": "It compiles but waits forever at runtime" }
            ],
        	"answer": {
            	"ids": ["F"],
            	"explanation": "The code compiles without issue, so options B, C, and D are incorrect. The limit on the cyclic barrier is 10, but the stream can generate only up to 9 threads that reach the barrier; therefore, the limit can never be reached, and option F is the correct answer, making options A and E incorrect. Even if the limit(9) statement was changed to limit(10), the program could still hang since the JVM might not allocate 10 threads to the parallel stream."
            }
        },
    	{
        	"number": 24,
        	"title": "What statements about the following class definition are true? (Choose all that apply.)",
        	"code": "public final class TicketManager {\n   private int tickets;\n   private static TicketManager instance;\n   private TicketManager() {}\n   static synchronized TicketManager getInstance() {      // k1\n      if (instance==null) instance = new TicketManager(); // k2\n      return instance;\n   }\n \n   public int getTicketCount() { return tickets; }\n   public void addTickets(int value) {tickets += value;}  // k3\n   public void sellTickets(int value) {\n      synchronized (this) {                               // k4\n         tickets -= value;\n      } } }",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "It compiles without issue" },
            	{ "id": "B", "text": "The code will not compile because of line k2" },
            	{ "id": "C", "text": "The code will not compile because of line k3" },
            	{ "id": "D", "text": "The locks acquired on k1 and k4 are on the same object" },
            	{ "id": "E", "text": "The class correctly protects the tickets data from race conditions" },
            	{ "id": "F", "text": "At most one instance of TicketManager will be created in an application that uses this class" }
            ],
        	"answer": {
            	"ids": ["F","A"],
            	"explanation": "The class compiles without issue, so option A is correct. Since getInstance() is a static method and sellTickets() is an instance method, lines k1 and k4 synchronize on different objects, making option D incorrect. The class is not thread- safe because the addTickets() method is not synchronized, and option E is incorrect. One thread could call sellTickets() while another thread calls addTickets(), possibly resulting in bad data. Finally, option F is correct because the getInstance() method is synchronized. Since the constructor is private, this method is the only way to create an instance of TicketManager outside the class. The first thread to enter the method will set the instance variable, and all other threads will use the existing value. This is a singleton pattern."
            }
        },
    	{
        	"number": 25,
        	"title": "Assuming an implementation of the performCount() method is provided prior to runtime, which of the following are possible results of executing the following application? (Choose all that apply.)",
        	"code": "import java.util.*;\nimport java.util.concurrent.*;\npublic class CountZooAnimals {\n   public static void performCount(int animal) {\n      // IMPLEMENTATION OMITTED   }\n   public static void printResults(Future<?> f) {\n      try {\n         System.out.println(f.get(1, TimeUnit.DAYS)); // o1\n      } catch (Exception e) {\n         System.out.println(\"Exception!\");\n      }\n   }\n   public static void main(String[] args) throws Exception {\n      final var r = new ArrayList<Future<?>>();\n      ExecutorService s = Executors.newSingleThreadExecutor();\n      try {\n         for(int i = 0; i < 10; i++) {\n            final int animal = i;\n            r.add(s.submit(() -> performCount(animal))); // o2\n         }\n         r.forEach(f -> printResults(f));\n      } finally { s.shutdown(); }\n   } }",
        	"multipleOptions": true,
        	"options": [
            	{ "id": "A", "text": "It outputs a number 10 times" },
            	{ "id": "B", "text": "It outputs a Boolean value 10 times" },
            	{ "id": "C", "text": "It outputs a null value 10 times" },
            	{ "id": "D", "text": "It outputs Exception! 10 times" },
            	{ "id": "E", "text": "It hangs indefinitely at runtime" },
            	{ "id": "F", "text": "The code will not compile because of line o1" },
            	{ "id": "G", "text": "The code will not compile because of line o2" }
            ],
        	"answer": {
            	"ids": ["D","C"],
            	"explanation": "The code compiles and runs without issue, so options F and G are incorrect. The return type of performCount() is void, so submit() is interpreted as being applied to a Runnable expression. While submit(Runnable) does return a Future<?>, calling get() on it always returns null. For this reason, options A and B are incorrect, and option C is correct. The performCount() method can also throw a runtime exception, which will then be thrown by the get() call as an ExecutionException; therefore, option D is also a correct answer. Finally, it is also possible for our performCount() to hang indefinitely, such as with a deadlock or infinite loop. Luckily, the call to get() includes a timeout value. While each call to Future.get() can wait up to a day for a result, it will eventually finish, so option E is incorrect."
            }
        }
    ]
}
